File: runtime/src/de/fatox/meta/MetaModule.java
package de.fatox.meta;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.PerspectiveCamera;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.graphics.g3d.utils.ModelBuilder;
import com.badlogic.gdx.utils.Json;
import de.fatox.meta.api.AssetProvider;
import de.fatox.meta.api.Logger;
import de.fatox.meta.api.entity.EntityManager;
import de.fatox.meta.api.graphics.FontProvider;
import de.fatox.meta.api.graphics.Renderer;
import de.fatox.meta.api.ui.UIManager;
import de.fatox.meta.api.ui.UIRenderer;
import de.fatox.meta.assets.MetaAssetProvider;
import de.fatox.meta.entity.Meta3DEntity;
import de.fatox.meta.entity.MetaEntityManager;
import de.fatox.meta.graphics.font.MetaFontProvider;
import de.fatox.meta.graphics.renderer.BufferRenderer;
import de.fatox.meta.injection.Log;
import de.fatox.meta.injection.Named;
import de.fatox.meta.injection.Provides;
import de.fatox.meta.injection.Singleton;
import de.fatox.meta.input.MetaInput;
import de.fatox.meta.sound.MetaSoundPlayer;
import de.fatox.meta.task.MetaTaskManager;
import de.fatox.meta.ui.MetaUIRenderer;
import de.fatox.meta.ui.MetaUiManager;

@Singleton
public class MetaModule {

    @Provides
    @Singleton
    public MetaSoundPlayer metaSoundPlayer() {
        return new MetaSoundPlayer();
    }

    @Provides
    @Singleton
    public UIRenderer uiRenderer(MetaUIRenderer metaUIRenderer) {
        return metaUIRenderer;
    }

    @Provides
    @Singleton
    public UIManager uiManager(MetaUiManager metaUiManager) {
        return metaUiManager;
    }

    @Provides
    @Singleton
    public MetaInput metaInput() {
        return new MetaInput();
    }

    @Provides
    @Singleton
    public Renderer renderer(BufferRenderer renderer) {
        return renderer;
    }

    @Provides
    @Singleton
    public PerspectiveCamera perspectiveCamera() {
        PerspectiveCamera cam = new PerspectiveCamera(67, Gdx.graphics.getWidth(), Gdx.graphics.getHeight());
        cam.position.set(0, 50f, 50f);
//        cam.lookAt(0, 0, 0);
        cam.near = 1;
        cam.far = 750f;
        return cam;
    }

    @Provides
    @Singleton
    public ModelBuilder modelBuilder() {
        return new ModelBuilder();
    }

    @Provides
    @Singleton
    @Named("default")
    public EntityManager<Meta3DEntity> entityManager() {
        return new MetaEntityManager();
    }


    @Provides
    @Singleton
    @Named("default")
    public SpriteBatch spriteBatch() {
        SpriteBatch spriteBatch = new SpriteBatch();
        spriteBatch.enableBlending();
        return spriteBatch;
    }

    @Provides
    @Singleton
    @Named("default")
    public AssetProvider assetProvider(MetaAssetProvider metaAssetProvider) {
        return metaAssetProvider;
    }

    @Provides
    @Singleton
    @Named("default")
    public FontProvider fontProvider(MetaFontProvider metaFontProvider) {
        return metaFontProvider;
    }


    @Provides
    @Singleton
    @Named("default")
    public MetaTaskManager taskManager() {
        return new MetaTaskManager();
    }

    @Provides
    @Singleton
    @Log
    public Logger log(MetaLogger metaLogger) {
        return metaLogger;
    }

    @Provides
    @Singleton
    @Named("default")
    public Json json() {
        return new Json();
    }

}

File: runtime/src/main/kotlin/de/fatox/meta/MetaModule.kt
package de.fatox.meta

import com.badlogic.gdx.Application
import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.PerspectiveCamera
import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.badlogic.gdx.graphics.g3d.utils.ModelBuilder
import com.badlogic.gdx.graphics.glutils.ShaderProgram
import com.badlogic.gdx.graphics.glutils.ShaderProgram.*
import com.badlogic.gdx.utils.Json
import de.fatox.meta.api.MetaInputProcessor
import de.fatox.meta.api.entity.EntityManager
import de.fatox.meta.api.graphics.FontProvider
import de.fatox.meta.api.ui.UIManager
import de.fatox.meta.api.ui.UIRenderer
import de.fatox.meta.entity.Meta3DEntity
import de.fatox.meta.entity.MetaEntityManager
import de.fatox.meta.graphics.font.FontInfo
import de.fatox.meta.graphics.font.MetaFontProvider
import de.fatox.meta.injection.MetaInject
import de.fatox.meta.input.MetaInput
import de.fatox.meta.sound.MetaSoundPlayer
import de.fatox.meta.task.MetaTaskManager
import de.fatox.meta.ui.MetaUIRenderer
import de.fatox.meta.ui.MetaUiManager
import de.fatox.meta.ui.UiControlHelper
import org.intellij.lang.annotations.Language

object MetaModule {
	init {
		MetaInject.global {
			singleton<FontProvider>("default") { MetaFontProvider() }
			singleton { MetaSoundPlayer() }
			singleton<UIRenderer> { MetaUIRenderer() }
			singleton<UIManager> { MetaUiManager() }
			singleton { ModelBuilder() }
			singleton<MetaInputProcessor> { MetaInput() }
			singleton<EntityManager<Meta3DEntity>>("default") { MetaEntityManager() }
			singleton("default") { MetaTaskManager() }
			singleton("", "default")
			singleton(Json(), "default")
			singleton("default") { UiControlHelper() }
			singleton("default") { FontInfo("Montserrat.ttf", "Montserrat-Bold.ttf","RobotoMono.ttf") }
			singleton("default") {
				SpriteBatch(1000, inject("spritebatch-shader")).apply { enableBlending() }
			}
			singleton {
				PerspectiveCamera(67f, Gdx.graphics.width.toFloat(), Gdx.graphics.height.toFloat()).apply {
					position.set(0f, 50f, 50f)
					near = 0.1f
					far = 200f
				}
			}
			singleton("spritebatch-shader") {
				@Suppress("SpellCheckingInspection")
				@Language("GLSL") val vertexShader =
					"""
in vec4 $POSITION_ATTRIBUTE;
in vec4 $COLOR_ATTRIBUTE;
in vec2 ${TEXCOORD_ATTRIBUTE}0;
uniform mat4 u_projTrans;
out vec4 v_color;
out vec2 v_texCoords;

void main()
{
   v_color = $COLOR_ATTRIBUTE;
   v_color.a = v_color.a * (255.0/254.0);
   v_texCoords = ${TEXCOORD_ATTRIBUTE}0;
   gl_Position =  u_projTrans * $POSITION_ATTRIBUTE;
}
""".trimMargin()

				@Suppress("SpellCheckingInspection")
				@Language("GLSL")
				val fragmentShader =
					"""
#ifdef GL_ES
#define LOWP lowp
precision mediump float;
#else
#define LOWP 
#endif
in LOWP vec4 v_color;
in vec2 v_texCoords;
uniform sampler2D u_texture;
out vec4 fragData;
void main()
{
	fragData = v_color * texture(u_texture, v_texCoords);
}
""".trimMargin()
				val shader = ShaderProgram(vertexShader, fragmentShader)
				require(shader.isCompiled) { "Error compiling shader: ${shader.log}" }
				shader
			}
		}
	}
}
